File: golang/src/detectors/integer-overflow/IntegerOverflow.go
package main

import (
	"fmt"
	"strconv"
)

// {fact rule=integer-overflow@v1.0 defects=1}
func integerOverflowNoncompliant() {
	value, err := strconv.Atoi("-2147483649")
	if err != nil {
		panic(err)
	}
	// Noncompliant: `int32` used with big value.
	newValue := int32(value)
	fmt.Println(newValue)
}
// {/fact}

// {fact rule=integer-overflow@v1.0 defects=0}
func integerOverflowCompliant() {
	value, err := strconv.Atoi("2147483647")
	if err != nil {
		panic(err)
	}
	// Compliant:  `int32` used with small value.
	newValue := int32(value)
	fmt.Println(newValue)
}
// {/fact}

File: csharp/src/detectors/integer-overflow/IntegerOverflow.cs
namespace OverFlow
{
    class IntegerOverflow
    {
        // {fact rule=integer-overflow@v1.0 defects=1}
        public void IntegerOverflowNoncompliant()
        {
            long data;
            data = long.MaxValue;
            // Noncompliant: if `data == long.MaxValue` then this will overflow.
            long result = (long)(data + 1);
            Console.WriteLine("result: " + result);
        }
        // {/fact}
        
        // {fact rule=integer-overflow@v1.0 defects=0}
        public void IntegerOverflowCompliant()
        {
            long data;
            // Compliant: Use a hardcoded number that won't cause overflow.
            data = 2;
            long result = (long)(data + 1);
            Console.WriteLine("result: " + result);
        }
        // {/fact}
    }
}

